### PR TITLE
Expose E2E build errors

### PR DESCRIPTION
**Pull request by tevanoff**
Currently, any error thrown during the build step of an E2E build is treated as a missing dependency error, and the error message output is very specific to that. However, the build can fail for other reasons, such as archive directory not found. These other legit errors are swallowed and the resulting output is confusing.

This flips things around to try to determine if the error is due to a missing dependency, where the error usually has something to do about a command not being found. If the error message doesn't match, this will output a more generic error message that includes the original error message. 

**How to test**
* With a local project, install the canary CLI and try to run the chromatic command in both of the following scenarios and see that the errors are correct:
  * With the E2E dep installed, but with the archive directory removed
  * With the archive directory in place, but without the E2E dep installed
* Using the canary github action (`chromaui/action-canary@v1`), test a project with the above two scenarios and see that the errors are correct. 

<details>
  <summary>📦 Published PR as canary version: <code>11.0.5--canary.940.8192597877.0</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install chromatic@11.0.5--canary.940.8192597877.0
  # or 
  yarn add chromatic@11.0.5--canary.940.8192597877.0
  ```
</details>

### CODE REVIEW

File: node-src/tasks/build.ts [hunk 1]
```diff
@@ -77,7 +77,7 @@ export const buildStorybook = async (ctx: Context) => {
     ctx.log.debug('Runtime metadata:', JSON.stringify(ctx.runtimeMetadata, null, 2));
 
     const subprocess = execaCommand(ctx.buildCommand, {
-      stdio: [null, logFile, logFile],
+      stdio: [null, logFile, null],
```
Comment (tevanoff) [inline node-src/tasks/build.ts:80]: This pipes stderr to the terminal rather than the storybook build log file, which seems better in general?

File: node-src/ui/messages/errors/e2eBuildFailed.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  flag,
+  errorMessage,
+}: {
+  flag: string;
+  errorMessage: string;
+}) => {
+  return dedent(chalk`
+      ${error} Failed to run \`chromatic --${flag}\`:
+      
+      ${errorMessage}
```
Comment (tevanoff) [inline node-src/ui/messages/errors/e2eBuildFailed.ts:15]: We could add a list of causes for this error, or maybe better yet link to a troubleshooting page in our docs.

File: node-src/tasks/build.ts [hunk 3]
```diff
@@ -87,15 +87,12 @@ export const buildStorybook = async (ctx: Context) => {
     // and it failed, that means we couldn't find it. This probably means they haven't
     // installed the right dependency or run from the right directory
     if (
-      ctx.options.inAction &&
-      (ctx.options.playwright || ctx.options.cypress) &&
-      e.message.match(e2EbuildBinName)
+      e.message.match(e2EbuildBinName) &&
```
Comment (tevanoff) [inline node-src/tasks/build.ts:90]: Not sure if we even need this check on the bin name. The command that failed is in the error message, so I don't know when this wouldn't match for E2E projects.

File: node-src/tasks/build.ts (hunk 1 above)
Comment (tmeasday) [reply node-src/tasks/build.ts:80]: I'm not sure what SB outputs to stderr vs stdout?

File: node-src/tasks/build.ts (hunk 3 above)
Comment (tmeasday) [reply node-src/tasks/build.ts:90]: Right, I guess it's probably rendundant.

File: node-src/ui/messages/errors/e2eBuildFailed.ts (hunk 2 above)
Comment (tmeasday) [reply node-src/ui/messages/errors/e2eBuildFailed.ts:15]: I guess the main one we are worried about is when an error like `no such binary exists "build-archive-storybook"`. For a user that's going to be a pretty opaque error if they have no idea about what archive SB even is.

File: node-src/tasks/build.ts [hunk 4]
```diff
@@ -60,6 +61,34 @@ export const setBuildCommand = async (ctx: Context) => {
...
+  return errorRegexes.some((regex) => msg.match(new RegExp(regex, 'gi')));
+}
+
+function e2eBuildErrorMessage(err, workingDir, ctx: Context): { exitCode: number, message: string } {
+  const flag = ctx.options.playwright ? 'playwright' : 'cypress';
+
+  // If we tried to run the E2E package's bin directly (due to being in the action)
+  // and it failed, that means we couldn't find it. This probably means they haven't
+  // installed the right dependency or run from the right directory.
+  if (isE2EBuildCommandNotFoundError(err)) {
```
Comment (tevanoff) [inline node-src/tasks/build.ts:78]: @tmeasday as per our slack thread, I fixed this to try to determine if it's the command not found error.

File: node-src/tasks/build.ts [hunk 5]
```diff
@@ -60,6 +61,34 @@ export const setBuildCommand = async (ctx: Context) => {
 const timeoutAfter = (ms) =>
   new Promise((resolve, reject) => setTimeout(reject, ms, new Error(`Operation timed out`)));
 
+function isE2EBuildCommandNotFoundError(errorMessage: string) {
+  // It's hard to know if this is the case as each package manager has a different type of
+  // error for this, but we'll try to figure it out.
+  const errorRegexes = ['command not found', `[\\W]?${e2eBuildBinName}[\\W]? not found`, 'code E404', 'exit code 127', `command failed.*${e2eBuildBinName}.*$`];
+  return errorRegexes.some((regex) => errorMessage.match(new RegExp(regex, 'gi')));
```
[the remark this answers is not in the document]
Comment (tmeasday) [reply node-src/tasks/build.ts:68]: This is decent advice (be careful about dynamic regexps) from Codacy but the solution doesn't make sense to me -- I could be off though. cc @paulelliott

File: node-src/tasks/build.ts [hunk 6]
```diff
@@ -60,6 +61,34 @@
...
```
[the remark this answers is not in the document]
Comment (tmeasday) [reply node-src/tasks/build.ts:68]: @paulelliott my feeling is the AI is getting really wordy here and I am glazing over and missing the key points:

1. Be careful about dynamic strings in regexps
2. You can comment away the warning
3. However, be super careful if you are going to do that.

File: node-src/tasks/build.test.ts [hunk 7]
```diff
@@ -166,3 +166,45 @@ describe('buildStorybook', () => {
     );
   });
 });
+
+describe('buildStorybook E2E', () => {
+  it('fails with missing dependency error when dependency not installed', async () => {
```
Comment (tmeasday) [inline node-src/tasks/build.test.ts:171]: you could use `test.each()` here?

**Review by tmeasday — APPROVED** [review]
I like it, thanks for working through it @tevanoff

File: node-src/tasks/build.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (paulelliott) [reply node-src/tasks/build.ts:68]: The other piece of advice it had was to use the RE2 library from google to parse user-supplied regex in a safe way. That feels like an easy change to make and right way to resolve these issues. The initial comment it put on the PR yesterday said that and I'm not sure why it shifted to this. You can see that advice on the second warning on the Codacy app.

https://app.codacy.com/gh/chromaui/chromatic-cli/pullRequest?prid=13786760

File: node-src/tasks/build.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (paulelliott) [reply node-src/tasks/build.ts:68]: When I enabled this I just blindly turned the whole thing on. It is quite configurable in the application so if there are rules we disagree with we can easily turn them off per repo. You can also dismiss this specific warning in their interface and it keeps a log of all the specific places it was ignored.

File: node-src/tasks/build.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (tevanoff) [reply node-src/tasks/build.ts:68]: This is good advice, however, the `e2eBuildBinName` that it is calling out is a simple string constant, not user-controlled input, so this is all moot.

[the remark this answers is not in the document]
Comment (tevanoff) [reply node-src/tasks/build.ts:68]: @tmeasday I can't make sense of that suggestion to use `.replace()` either 🤔

File: node-src/lib/setExitCode.ts [hunk 8]
```diff
@@ -25,6 +25,9 @@ export const exitCodes = {
   STORYBOOK_START_FAILED: 22,
   STORYBOOK_BROKEN: 23,
 
+  // E2E errors
+  E2E_BUILD_FAILED: 51,
+
```
Comment (skitterm) [inline node-src/lib/setExitCode.ts:30]: Should we add this to the CLI [exit codes](https://www.chromatic.com/docs/cli/#exit-codes) in the documentation?

File: node-src/tasks/build.ts [hunk 9]
```diff
@@ -60,6 +61,40 @@ export const setBuildCommand = async (ctx: Context) => {
...
 
+function isE2EBuildCommandNotFoundError(errorMessage: string) {
+  // It's hard to know if this is the case as each package manager has a different type of
+  // error for this, but we'll try to figure it out.
+  const errorRegexes = [
+    'command not found', // `Command not found: build-archive-storybook`
+    `[\\W]?${e2eBuildBinName}[\\W]? not found`, // `Command "build-archive-storybook" not found`
+    'code E404', // npm not found error can include this code
+    'exit code 127', // Exit code 127 is a generic not found exit code
+    `command failed.*${e2eBuildBinName}.*$`]; // A single line error from execa like `Command failed: yarn build-archive-storybook ...`
```
Comment (skitterm) [inline node-src/tasks/build.ts:72]: Thanks for doc-ing the edge-cases and where each comes from.

File: node-src/tasks/build.test.ts [hunk 10]
```diff
@@ -166,3 +166,47 @@ describe('buildStorybook', () => {
...
+
+  it('fails with generic error message when not missing dependency error', async () => {
+    const ctx = {
+      buildCommand: 'npm exec build-archive-storybook',
+      options: { buildScriptName: '', playwright: true },
+      env: { STORYBOOK_BUILD_TIMEOUT: 0 },
+      log: { debug: vi.fn(), error: vi.fn() },
+    } as any;
+
+    const errorMessage = 'Command failed with exit code 1: npm exec build-archive-storybook --output-dir /tmp/chromatic--4210-0cyodqfYZabe\n\nMore error message lines\n\nAnd more';
```
Comment (skitterm) [inline node-src/tasks/build.test.ts:205]: @tevanoff this generic error message includes (for example) that the archive directory isn't found? Just wanted to make sure we're still passing that part through.

Comment (tevanoff) [reply node-src/tasks/build.test.ts:205]: @skitterm yes, this is testing that the generic error message (`Command failed with exit code 1: npm exec build-archive-storybook...`) that has multiple lines gets output using our the generic E2E build error template, which will out put the full error. That's how things like the archive directory not found errors are shaped.

File: node-src/lib/setExitCode.ts (hunk 8 above)
Comment (tevanoff) [reply node-src/lib/setExitCode.ts:30]: @skitterm sure, I can open a PR for that.

**Review by skitterm — APPROVED** [review]

**Comment by ghengeveld** [comment]
:rocket: PR was released in [`v11.0.6`](https://github.com/chromaui/chromatic-cli/releases/tag/v11.0.6) :rocket:
